### PR TITLE
[Requests for Contributions (Issues)] Added test for a bug about regularization.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
## The issue
I added a test showing that the regularization has a different behavior if a layer is shared or if a model that has this layer is shared. 

If a layer is shared, the regularization is counted only once, but if a model is shared, the regularization corresponding to this model is counted multiple times (as many times as this shared model is used).

It was show in this issue: #8807 

## The test
In the test, I compared the L1 regularization loss of two models. The first one has a shared dense layer, the second one has a shared dense layer, but this layer is wrapped in another model.

I don't know if this is an expected behavior or not, but it sure is strange. If this is a bug, let me know so that I can work on it. 

Thank you.

### CODE REVIEW

File: tests/keras/regularizers_test.py [hunk 1]
```diff
@@ -56,5 +58,32 @@ def test_activity_regularization():
...
+        out2 = func(input_2)
+        out = Average()([out1, out2])
+        model = Model([input_1, input_2], out)
+        model.compile(optimizer='sgd', loss='mse', metrics=['mse'])
+        model.fit([dummy, dummy], dummy, verbose=1, epochs=1)
+        hist = model.history.history
+        print(hist)
+        l1_loss.append(hist['loss'][0] - hist['mean_squared_error'][0])
+
+    assert assert_allclose(l1_loss[0], l1_loss[1], atol=1e-4)()
```
Comment (fchollet) [inline tests/keras/regularizers_test.py]: This is not the way to assert whether regularization is being applied. Instead, look at the contents of `model.losses`.

**Comment by fchollet** [comment]
I think this may be a bug, but the test doesn't show it clearly. Please update the test.

**Comment by gabrieldemarmiesse** [comment]
Thank you for the feedback!  I will make this more clear. 
 I'll also try to dig and see where the bug comes from too.

**Comment by gabrieldemarmiesse** [comment]
Hi!
So I modified the tests based in your suggestion. 
I also made 3 different functions because I think that there are 3 main cases to be tested.
Every time we verify that `len(model.losses == 1)`

## test_regularization_shared_layer() -> PASSED
The same layer is used twice in the final model.

## test_regularization_shared_model() -> FAILED
The same model is used twice in the final model

## test_regularization_shared_layer_in_different_models() -> FAILED
Two different model instances are used in the final model. 
But those two model instances refer to a single unique layer.

Now that we have proper testing functions, I'll try to fix the bug and commit on this branch.

**Comment by gabrieldemarmiesse** [comment]
So for the heads up, after reading the code, I noticed that a model ask each sub model for it's `self._per_input_losses`. But the issue is that there is no check at all on where the loss comes from. So if it came from the same weight, the model will add twice the tensor to the list of losses. We could do bookeeping but since there seem to be a unique identifier for each tensor, maybe we could use this.

Any idea is welcome.

**Comment by gabrieldemarmiesse** [comment]
Hi, 

To sum things up:

## The bug:
In a model that uses multiple times a layer, if the layer in "hidden" inside other models, then the regularization get counted multiple times. The best test to put this in evidence is this one:

```
def create_multi_input_model_from(layer1, layer2):
    input_1 = Input(shape=(data_dim,))
    input_2 = Input(shape=(data_dim,))
    out1 = layer1(input_1)
    out2 = layer2(input_2)
    out = Average()([out1, out2])
    model = Model([input_1, input_2], out)
    return model

def test_regularization_shared_layer_in_different_models():
    (x_train, y_train), (x_test, y_test) = get_data()
    dense_layer = Dense(num_classes,
                        kernel_regularizer=regularizers.l1())
    models = []
    for _ in range(2):
        input_tensor = Input(shape=(data_dim,))
        models.append(Model(input_tensor, dense_layer(input_tensor)))

    model = create_multi_input_model_from(*models)
    model.compile(loss='categorical_crossentropy', optimizer='sgd')
    assert len(model.losses) == 1
```

## Why does this happen?

When using the attribute `losses` or using the functions `add_losses` or `get_losses_for`, lists of tensors are concatenated but there is no check on the provenance of the tensors, so two references of a same tensor could end up in the same resulting list. Causing the final list `model.losses` to have multiple references to the same tensor representing regularization loss.

## The fix:

Every time lists of unconditional losses are concatenated, instead of using the `+` operator, we use the `Layer._concat_unique(list1, list2)` which performs a check with the function `id()` to ensure unicity.

## The tests:
There are now 4 added test functions:

```
test_concat_unique()
test_regularization_shared_layer()
test_regularization_shared_model()
test_regularization_shared_layer_in_different_models()
```

## Dark corners:

- I hope I found all the places where lists of unconditional losses are concatenated.
- If there are integers or floats appearing in the `_concat_unique` function, we don't check and add them to the resulting list unconditionally.
- For some misterious reason, only the CNTK tests are failing. Giving the error `ValueError: Learner's parameters list must not contain duplicates.`
- I hope the same bug doesn't appear with unconditional updates. Though I'm not familiar with updates.

**Comment by fchollet** [comment]
I haven't looked at the issue and at potential fixes in detail, but I think the fix is potentially bug-prone because of the use of `id`. This would only work with symbolic tensors.

Maybe in `def losses(self)` we can keep track of the origin of the losses, and de-dupe at that stage.

**Comment by gabrieldemarmiesse** [comment]
Hi! Thanks for the advice. 
I don't mind doing some bookkeeping (I hope it won't bloat the code too much).

I have a question though, how can I determine the origin of the loss when the user does this:

```python
a = Input(shape=(3,), name='input_a')
a_2 = Dense(4, name='dense_1')(a)
a_3 = Dense(4, name='dense_2')(a)

model = Model(a, [a_2, a_3])
model.add_loss(K.mean(a_3 + a_2))
```

Or this case with integers:

```python
a = Input(shape=(3,), name='input_a')
a_2 = Dense(4, name='dense_1')(a)
a_3 = Dense(4, name='dense_2')(a)

model = Model(a, [a_2, a_3])
model.add_loss(1, None)
```


whithout changing the public API and adding breaking changes, I don't know how I can check where does the tensor/ integer/float come from in those cases (I took those examples from the keras test suite so I assumed that the users could do it too).

**Pull request closed without merging**